src/lib/core/sample_sort.c

```diff
@@ -0,0 +1,513 @@
+		assert(m >= 0 && m < sort->threads_num - 1);
+		if (sort->cmp(elem, sort->splitters + m * sort->elem_size,
+			      sort->cmp_arg) < 0)
+			e = m;
+		else
+			b = m;
+	} while (e - b > 1);
+	return b + 1;
+}
+
+/**
+ * Calculate element to bucket map for data part assigned to a thread.
+ * Additionally calculate bucket histogramm - how much elements are placed
+ * in each bucket.
+ */
+static int
+sample_sort_find_oracles(va_list ap)
+{
+	struct sample_sort_worker *worker = va_arg(ap, typeof(worker));
+	struct sample_sort *sort = worker->sort;
+
+	void *pos = sort->data + worker->begin * sort->elem_size;
+	for (size_t i = worker->begin; i < worker->end; i++) {
+		int b = sample_sort_find_bucket(sort, pos);
+		assert(b >= 0 && b < sort->threads_num);
+		sort->oracles[i] = b;
+		worker->bucket_hist[b]++;
+		pos += sort->elem_size;
+	}
+
+	return 0;
+}
+
+/**
+ * Distribute data part assigned to a thread to buckets. Each bucket
+ * has a designated place for this thread.
+ */
+static int
+sample_sort_split_to_buckets(va_list ap)
+{
+	struct sample_sort_worker *worker = va_arg(ap, typeof(worker));
+	struct sample_sort *sort = worker->sort;
+
+	void *pos = sort->data + worker->begin * sort->elem_size;
+	for (size_t i = worker->begin; i < worker->end; i++) {
+		int b = sort->oracles[i];
+		memcpy(sort->buffer + worker->bucket_offs[b], pos,
+		       sort->elem_size);
+		worker->bucket_offs[b] += sort->elem_size;
+		pos += sort->elem_size;
+	}
+
+	return 0;
+}
+
+/**
+ * Sort bucket assigned to a thread and copy sorted data back to the original
+ * array.
+ */
+static int
+sample_sort_bucket_sort(va_list ap)
+{
+	struct sample_sort_worker *worker = va_arg(ap, typeof(worker));
+	struct sample_sort *sort = worker->sort;
+
+	/* Sort this worker bucket. */
+	qsort_arg_st(sort->buffer + worker->bucket_begin * sort->elem_size,
+		     worker->bucket_size, sort->elem_size,
+		     sort->cmp, sort->cmp_arg);
+
+	/* Move sorted data back from temporary space. */
+	memcpy(sort->data + worker->bucket_begin * sort->elem_size,
+	       sort->buffer + worker->bucket_begin * sort->elem_size,
+	       worker->bucket_size * sort->elem_size);
+
+	return 0;
+}
+
+/**
+ * Run function in several threads. Yields while waiting threads to
+ * finish.
+ *
+ * Arguments:
+ *  func        - a function to run in threads
+ *  workers     - array of function arguments. An element of this array will
+ *                be passed to as an argument to the function `func` for each
+ *                thread. So the array should have `threads_num` elements
+ *  threads_num - number of threads
+ */
+static void
+sample_sort_corun_mt(fiber_func func,
+		     struct sample_sort_worker *workers, int threads_num)
+{
+	struct cord *cords = xmalloc(threads_num * sizeof(struct cord));
+
+	for (int i = 0; i < threads_num; i++) {
+		char name[FIBER_NAME_MAX];
+
+		snprintf(name, sizeof(name), "sort.worker.%d", i);
+		if (cord_costart(&cords[i], name, func, &workers[i]) != 0) {
```

Comment:
I don't think you need the event loop in worker threads, i.e. you should use `cord_start`.

Reply:
Ufortunately `cord_start`/`cord_cojoin` does not work. The latter fails on [assertion](https://github.com/tarantool/tarantool/blob/199e20b9fbfa97bc398abbfe0621b69e2b1e80d2/src/lib/core/fiber.c#L1808) in `cord_join`:
```c
if (f->f_ret != 0) {
	assert(!diag_is_empty(&f->diag));
	diag_move(&f->diag, diag_get());
	res = -1;
}
```